{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example of running tool \n",
    "\n",
    "First, let's load the PredictInteractions class that will make the predictions. Please make sure that you have the conda environment setup correctly. The list of modules required to run this tool can be found in the `requirements.txt` file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from vip.predict_interactions import PredictInteractions\n",
    "\n",
    "# path of saved machine learning model \n",
    "model_path = './vip/gbrt.pkl'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, the user need to define certain parameters:\n",
    "1. Location of viruses fasta files. All viruses should be in their own separate file. \n",
    "2. Location of the host fasta files. All hosts should be in their own separate files as well. \n",
    "3. Blastn results of viruses against host, and viruses against spacers. \n",
    "4. The filename for the output. \n",
    "4. The number of CPU cores to be used. \n",
    "\n",
    "IMPORTANT: Viruses and hosts have to be in separate folders. \n",
    "\n",
    "In here, I will be using the test_set folder as an example. \n",
    "\n",
    "There are two different ways to use the tool: \n",
    "1. Make a prediction for each possible virus-host pair. \n",
    "2. Make predictions only for virus-host pair of interest by providing a tsv as an additional input file. \n",
    "\n",
    "In either case, user need to provide the virus and host sequences in separate folders, and output of blastn. \n",
    "\n",
    "### 1. **Make a prediction for each possible virus-host pair**\n",
    "\n",
    "\n",
    "Regarding the number of CPU cores, I strongly recommend to use at least 6 if that's possible. With only 1 core, it takes around 40 minutes to run this tool on the test set. Using 6, it will only take about 5 minutes. \n",
    "\n",
    "This tool also assumes that the user is interesting in predicting interactions for every virus-host combinations possible. \n",
    "\n",
    "The output file will be saved in the `test_set/` folder.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# USER INPUTS\n",
    "virus_directory_path = './test_set/virus_sequences/'\n",
    "host_directory_path = './test_set/host_sequences/'\n",
    "\n",
    "blastn_path = './test_set/StaphStudy_virusvhosts.tsv'\n",
    "spacer_path = './test_set/StaphStudy_virusvspacers_blastn.tsv'\n",
    "\n",
    "output_filename = './test_set/test_predictions.tsv'\n",
    "\n",
    "CPU_CORES = 6"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The code below computes the predictions. There is nothing for the user to change!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SETUP - ...indexing fasta filenames for viruses and hosts...\n",
      "SETUP - ...initialize all pairs...\n",
      "None\n",
      "-------> There are 40 viral sequences\n",
      "-------> There are 27 host sequences\n",
      "-------> Total number of interactions: 1080\n",
      "SETUP - ...getting fasta headers...\n",
      "SETUP - ...process blastn and spacers output...\n",
      "SETUP - ...calculate GC content and k-mer profiles...\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_pseudintermedius.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_hominis.fasta\n",
      "-------> current pair: GCA_021090845.1_ASM2109084v1_genomic.fasta | Staphylococcus_pettenkoferi.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_pseudintermedius.fasta\n",
      "-------> current pair: GCA_021090955.1_ASM2109095v1_genomic.fasta | Staphylococcus_pettenkoferi.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_hominis.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_fleuretti.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_kloosii.fasta\n",
      "-------> current pair: GCA_021090845.1_ASM2109084v1_genomic.fasta | Staphylococcus_saprophyticus.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_fleuretti.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_warneri.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_warneri.fasta\n",
      "-------> current pair: GCA_021090955.1_ASM2109095v1_genomic.fasta | Staphylococcus_saprophyticus.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_kloosii.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_capitis.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_arlettae.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_capitis.fasta\n",
      "-------> current pair: GCA_021090845.1_ASM2109084v1_genomic.fasta | Staphylococcus_vitulinus.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_arlettae.fasta\n",
      "-------> current pair: GCA_021090955.1_ASM2109095v1_genomic.fasta | Staphylococcus_vitulinus.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_succinus.fasta\n",
      "-------> current pair: GCA_021090845.1_ASM2109084v1_genomic.fasta | Staphylococcus_nepalensis.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_succinus.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_xylosus.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_xylosus.fasta\n",
      "-------> current pair: GCA_021090955.1_ASM2109095v1_genomic.fasta | Staphylococcus_nepalensis.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_hyicus.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_hyicus.fasta\n",
      "-------> current pair: GCA_021090845.1_ASM2109084v1_genomic.fasta | Staphylococcus_auricularis.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_sciuri.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_sciuri.fasta\n",
      "-------> current pair: GCA_021090955.1_ASM2109095v1_genomic.fasta | Staphylococcus_auricularis.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_pasteuri.fasta\n",
      "-------> current pair: GCA_021090845.1_ASM2109084v1_genomic.fasta | Staphylococcus_caprae.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_pasteuri.fasta\n",
      "-------> current pair: GCA_021090955.1_ASM2109095v1_genomic.fasta | Staphylococcus_caprae.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_epidermidis.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_epidermidis.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_lugdunensis.fasta\n",
      "-------> current pair: GCA_021090845.1_ASM2109084v1_genomic.fasta | Staphylococcus_cohnii.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_lugdunensis.fasta\n",
      "-------> current pair: GCA_021090955.1_ASM2109095v1_genomic.fasta | Staphylococcus_cohnii.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_aureus.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_aureus.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_lentus.fasta\n",
      "-------> current pair: GCA_021090845.1_ASM2109084v1_genomic.fasta | Staphylococcus_schleiferi.fasta\n",
      "-------> current pair: GCA_021090955.1_ASM2109095v1_genomic.fasta | Staphylococcus_schleiferi.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_lentus.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_equorum.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_equorum.fasta\n",
      "-------> current pair: GCA_021090845.1_ASM2109084v1_genomic.fasta | Staphylococcus_haemolyticus.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_chromogenes.fasta\n",
      "-------> current pair: GCA_021090955.1_ASM2109095v1_genomic.fasta | Staphylococcus_haemolyticus.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_chromogenes.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_pettenkoferi.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_pettenkoferi.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_hominis.fasta\n",
      "-------> current pair: GCA_021090935.1_ASM2109093v1_genomic.fasta | Staphylococcus_pseudintermedius.fasta\n",
      "-------> current pair: GCA_021090855.1_ASM2109085v1_genomic.fasta | Staphylococcus_pseudintermedius.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_hominis.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_saprophyticus.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_saprophyticus.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_fleuretti.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_kloosii.fasta\n",
      "-------> current pair: GCA_021090935.1_ASM2109093v1_genomic.fasta | Staphylococcus_warneri.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_fleuretti.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_kloosii.fasta\n",
      "-------> current pair: GCA_021090855.1_ASM2109085v1_genomic.fasta | Staphylococcus_warneri.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_vitulinus.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_vitulinus.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_arlettae.fasta\n",
      "-------> current pair: GCA_021090935.1_ASM2109093v1_genomic.fasta | Staphylococcus_capitis.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_arlettae.fasta\n",
      "-------> current pair: GCA_021090855.1_ASM2109085v1_genomic.fasta | Staphylococcus_capitis.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_nepalensis.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_nepalensis.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_xylosus.fasta\n",
      "-------> current pair: GCA_021090935.1_ASM2109093v1_genomic.fasta | Staphylococcus_succinus.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_xylosus.fasta\n",
      "-------> current pair: GCA_021090855.1_ASM2109085v1_genomic.fasta | Staphylococcus_succinus.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_auricularis.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_auricularis.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_sciuri.fasta\n",
      "-------> current pair: GCA_021090935.1_ASM2109093v1_genomic.fasta | Staphylococcus_hyicus.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_caprae.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_sciuri.fasta\n",
      "-------> current pair: GCA_021090855.1_ASM2109085v1_genomic.fasta | Staphylococcus_hyicus.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_caprae.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_epidermidis.fasta\n",
      "-------> current pair: GCA_021090935.1_ASM2109093v1_genomic.fasta | Staphylococcus_pasteuri.fasta\n",
      "-------> current pair: GCA_021090855.1_ASM2109085v1_genomic.fasta | Staphylococcus_pasteuri.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_cohnii.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_epidermidis.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_cohnii.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_aureus.fasta\n",
      "-------> current pair: GCA_021090935.1_ASM2109093v1_genomic.fasta | Staphylococcus_lugdunensis.fasta\n",
      "-------> current pair: GCA_021090855.1_ASM2109085v1_genomic.fasta | Staphylococcus_lugdunensis.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_schleiferi.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_aureus.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_schleiferi.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_equorum.fasta\n",
      "-------> current pair: GCA_021090935.1_ASM2109093v1_genomic.fasta | Staphylococcus_lentus.fasta\n",
      "-------> current pair: GCA_021090975.1_ASM2109097v1_genomic.fasta | Staphylococcus_haemolyticus.fasta\n",
      "-------> current pair: GCA_021090855.1_ASM2109085v1_genomic.fasta | Staphylococcus_lentus.fasta\n",
      "-------> current pair: GCA_021090825.1_ASM2109082v1_genomic.fasta | Staphylococcus_equorum.fasta\n",
      "-------> current pair: GCA_021090835.1_ASM2109083v1_genomic.fasta | Staphylococcus_haemolyticus.fasta\n",
      "-------> current pair: GCA_021090945.1_ASM2109094v1_genomic.fasta | Staphylococcus_pettenkoferi.fasta\n",
      "-------> current pair: GCA_021090935.1_ASM2109093v1_genomic.fasta | Staphylococcus_chromogenes.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process SpawnPoolWorker-5:\n",
      "Process SpawnPoolWorker-1:\n",
      "Process SpawnPoolWorker-6:\n",
      "Process SpawnPoolWorker-4:\n",
      "Process SpawnPoolWorker-3:\n",
      "Process SpawnPoolWorker-2:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 48, in mapstar\n",
      "    return list(map(*args))\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/compute_ml_features.py\", line 316, in compute_feature\n",
      "    k6distance.distance()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line 109, in distance\n",
      "    self.nucleotide_count()\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line -1, in nucleotide_count\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 48, in mapstar\n",
      "    return list(map(*args))\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/compute_ml_features.py\", line 312, in compute_feature\n",
      "    k3distance.distance()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line 109, in distance\n",
      "    self.nucleotide_count()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line -1, in nucleotide_count\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 48, in mapstar\n",
      "    return list(map(*args))\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/compute_ml_features.py\", line 316, in compute_feature\n",
      "    k6distance.distance()\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line 109, in distance\n",
      "    self.nucleotide_count()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line -1, in nucleotide_count\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 48, in mapstar\n",
      "    return list(map(*args))\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/compute_ml_features.py\", line 316, in compute_feature\n",
      "    k6distance.distance()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line 109, in distance\n",
      "    self.nucleotide_count()\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line -1, in nucleotide_count\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 48, in mapstar\n",
      "    return list(map(*args))\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/compute_ml_features.py\", line 316, in compute_feature\n",
      "    k6distance.distance()\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line 109, in distance\n",
      "    self.nucleotide_count()\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line -1, in nucleotide_count\n",
      "  File \"/Users/gbastien/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py\", line 48, in mapstar\n",
      "    return list(map(*args))\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/compute_ml_features.py\", line 312, in compute_feature\n",
      "    k3distance.distance()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line 110, in distance\n",
      "    self.null()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line 146, in null\n",
      "    self.nucleotide_count()\n",
      "  File \"/Users/gbastien/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/genomes_features.py\", line -1, in nucleotide_count\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m predictions\u001b[39m.\u001b[39mload_model(model_path)\n\u001b[1;32m      5\u001b[0m predictions\u001b[39m.\u001b[39mdo_setup()\n\u001b[0;32m----> 6\u001b[0m predictions\u001b[39m.\u001b[39;49mrun_parallel(CPU_CORES)\n\u001b[1;32m      7\u001b[0m predictions\u001b[39m.\u001b[39mpredict()\n\u001b[1;32m      8\u001b[0m predictions\u001b[39m.\u001b[39msave_predictions(output_filename)\n",
      "File \u001b[0;32m~/Documents/Research/Projects/VirusInfectionPredictor/vip/mlmodel/compute_ml_features.py:294\u001b[0m, in \u001b[0;36mComputeFeatures.run_parallel\u001b[0;34m(self, num_procs)\u001b[0m\n\u001b[1;32m    285\u001b[0m \u001b[39m'''\u001b[39;00m\n\u001b[1;32m    286\u001b[0m \u001b[39mCall the compute_feature method to compute features for each virus-host pair. This step is parallelized \u001b[39;00m\n\u001b[1;32m    287\u001b[0m \u001b[39mto significantly improve run time. \u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    290\u001b[0m \u001b[39m:type num_procs: int\u001b[39;00m\n\u001b[1;32m    291\u001b[0m \u001b[39m'''\u001b[39;00m\n\u001b[1;32m    293\u001b[0m \u001b[39mwith\u001b[39;00m Pool(num_procs) \u001b[39mas\u001b[39;00m pool:\n\u001b[0;32m--> 294\u001b[0m     results \u001b[39m=\u001b[39m pool\u001b[39m.\u001b[39;49mmap(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcompute_feature, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mpairs)\n\u001b[1;32m    296\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcomputed_pairs \u001b[39m=\u001b[39m results\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py:367\u001b[0m, in \u001b[0;36mPool.map\u001b[0;34m(self, func, iterable, chunksize)\u001b[0m\n\u001b[1;32m    362\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mmap\u001b[39m(\u001b[39mself\u001b[39m, func, iterable, chunksize\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[1;32m    363\u001b[0m     \u001b[39m'''\u001b[39;00m\n\u001b[1;32m    364\u001b[0m \u001b[39m    Apply `func` to each element in `iterable`, collecting the results\u001b[39;00m\n\u001b[1;32m    365\u001b[0m \u001b[39m    in a list that is returned.\u001b[39;00m\n\u001b[1;32m    366\u001b[0m \u001b[39m    '''\u001b[39;00m\n\u001b[0;32m--> 367\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_map_async(func, iterable, mapstar, chunksize)\u001b[39m.\u001b[39;49mget()\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py:768\u001b[0m, in \u001b[0;36mApplyResult.get\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    767\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget\u001b[39m(\u001b[39mself\u001b[39m, timeout\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m--> 768\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mwait(timeout)\n\u001b[1;32m    769\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mready():\n\u001b[1;32m    770\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mTimeoutError\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/vip/lib/python3.10/multiprocessing/pool.py:765\u001b[0m, in \u001b[0;36mApplyResult.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    764\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mwait\u001b[39m(\u001b[39mself\u001b[39m, timeout\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m--> 765\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_event\u001b[39m.\u001b[39;49mwait(timeout)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/vip/lib/python3.10/threading.py:607\u001b[0m, in \u001b[0;36mEvent.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    605\u001b[0m signaled \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_flag\n\u001b[1;32m    606\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m signaled:\n\u001b[0;32m--> 607\u001b[0m     signaled \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_cond\u001b[39m.\u001b[39;49mwait(timeout)\n\u001b[1;32m    608\u001b[0m \u001b[39mreturn\u001b[39;00m signaled\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/vip/lib/python3.10/threading.py:320\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    318\u001b[0m \u001b[39mtry\u001b[39;00m:    \u001b[39m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[39;00m\n\u001b[1;32m    319\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m--> 320\u001b[0m         waiter\u001b[39m.\u001b[39;49macquire()\n\u001b[1;32m    321\u001b[0m         gotit \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m    322\u001b[0m     \u001b[39melse\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# run model \n",
    "predictions = PredictInteractions(virus_directory_path, host_directory_path)\n",
    "predictions.add_blastn_files(blastn_path, spacer_path)\n",
    "predictions.load_model(model_path)\n",
    "predictions.do_setup()\n",
    "predictions.run_parallel(CPU_CORES)\n",
    "predictions.predict()\n",
    "predictions.save_predictions(output_filename)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Predict only for virus-host pair of interest**\n",
    "\n",
    "For this example, I will use the same set. But this I am also going to provide the additional parameter needed: a tsv file containing specific virus-host pair to be tested. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_of_interest_path = 'tests/datatests/example_VirusHost_pairlist.txt'\n",
    "\n",
    "\n",
    "output_filename = './test_set/test_onlydesiredpairs_predictions.tsv'\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With this parameter defined, now we can make the predictions!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SETUP - ...indexing fasta filenames for viruses and hosts...\n",
      "SETUP - ...initialize all pairs...\n",
      "tests/datatests/example_VirusHost_pairlist.txt\n",
      "reading pairs file\n",
      "Warning: file pair GCA_021090675.1_ASM2109067v1_genomic.fasta and Staphylococcus_epidermidis_noneexist.fasta are not all present. Please double check.\n",
      "SETUP - ...getting fasta headers...\n",
      "SETUP - ...process blastn and spacers output...\n",
      "SETUP - ...calculate GC content and k-mer profiles...\n",
      "-------> current pair: GCA_021090665.1_ASM2109066v1_genomic.fasta | Staphylococcus_aureus.fasta\n",
      "-------> current pair: GCA_021090665.1_ASM2109066v1_genomic.fasta | Staphylococcus_arlettae.fasta\n",
      "-------> current pair: GCA_021090675.1_ASM2109067v1_genomic.fasta | Staphylococcus_epidermidis.fasta\n",
      "MODEL - ...making predictions...\n",
      "MODEL - ...predictions are done!...\n"
     ]
    }
   ],
   "source": [
    "predictions = PredictInteractions(virus_directory_path, host_directory_path, pairs_of_interest=pairs_of_interest_path)\n",
    "predictions.add_blastn_files(blastn_path, spacer_path)\n",
    "predictions.load_model(model_path)\n",
    "predictions.do_setup()\n",
    "predictions.run_parallel(CPU_CORES)\n",
    "predictions.predict()\n",
    "predictions.save_predictions(output_filename)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vip",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
